Added docker build and quick-start

mgaffigan opened the pull request:
While this might be somewhat responsive to #40, it is primarily just to provide a quick-start way for a developer to build and test.  While trying to address Issue 5 I had hours of challenges getting things building happily - and shifted to docker as an escape hatch.

I do not expect this PR is ready to merge as is, but wanted to get something posted publicly.  Known issues include:

- [x] Not responsive to the acceptance criteria listed in #40
- [x] Uses a [personal container registry namespace](https://github.com/users/mgaffigan/packages/container/package/oie), rather than https://hub.docker.com/u/openintegrationengine
- [x] I'm not confident that I've implemented vmoptions handling correctly
- [x] Missing examples for docker-compose and playwithdocker

That said, it does work to the point of passing tests, running server, client, and simple message processing.

I'm new to the project, so if I've missed a step please let me know.

mgaffigan commented:
@tonygermano, I'm no longer aware of issues outside of the registry name (presumably someone owns the official URL?).

Differences I'm aware of relative to the original repo:

1. Uses multi-stage build to allow forks and keep dev, test, and prod on the same build pipeline and artifacts
2. Still unclear on jar signing story
3. Omitted [tests](https://github.com/nextgenhealthcare/connect-docker/tree/master/test) since they do not seem to run on any modern version of python and need to be reworked
4. Publish file is only pushing Temurin JDK on Ubuntu as ${tag}-jdk, and Temurin JRE on Alpine as ${tag} (e.g.: latest-jdk, latest-jre)

Happy to discuss as needed.  Every compose file, kubernetes deployment, etc... that I've tested doesn't notice the difference between this and stock.

tonygermano commented:
@mgaffigan  I pulled your launcher script out of this PR, made significant changes, and put it in its own PR in #119 . I think it is useful outside of a docker context, and it can still be used with docker. Would you mind reviewing it and giving any feedback?

mgaffigan commented:
@tonygermano, Excellent!  I'd love to see that merged and agree it is useful outside of docker.

mgaffigan commented:
@jonbartels,

Thanks for the review.  I think this PR got out of date and looks like it ended up with an accidental merge from #114 at some point.  Let me clean it up to minimums.  I'll pull the docker-compose examples out as well, assuming they fit better with #126.

mgaffigan commented on docker/mirth-connect.sh:
This script was the basis for #119.  It should not be duplicated for docker.  I'll update to assume #119 gets merged before this PR.

mgaffigan commented on Dockerfile:
```diff
@@ -0,0 +1,87 @@
+    && ANT_OPTS="-Dfile.encoding=UTF8" ant -f mirth-build.xml -DdisableSigning=true
+
+# Stage 2b: JDK runtime container
+FROM eclipse-temurin:21.0.7_6-jdk-noble as jdk-run
```
It can, but I disagree with it philosophically since then a particular git rev is not guaranteed to produce the same behavior.  Goal is to eliminate "it works on my machine".  I'd suggest something like [Dependbot](https://dev.to/oracle2025/how-to-keep-a-dockerfile-updated-with-dependabot-1mdn) be used to submit a PR when new versions are available.

If the project model is to "use latest", let me know and I will update it.

mgaffigan commented on DEVELOPERS.md:
```diff
@@ -0,0 +1,33 @@
+
+From the `server/` directory, run `ant -f mirth-build.xml -DdisableSigning=true`.
+
+If you are using Mirth Connect Administrator Launcher, you may need to omit
```
Updated.  Feel free to send suggested language if I didn't get it right.

mgaffigan replied:
Per discussion in Github, I updated to refer to the docs site, and only kept a reference to Ballista.

kpalang commented on .dockerignore:
suggestion: I think the "ignore everything but not this" approach would be more readable here.

For example:
```
*
!setup/
```

mgaffigan replied:
Closing as duplicate.

kpalang commented on Dockerfile:
I would split the Dockerfiles into two separate ones depending on where they source the `engine/` directory.

Firstly, I quite dislike build containers as I've had too many issues with layer caching platform-specific optimizations. With a long-building application like OIE it starts to matter. I'm not totally against it, but in my opinion it shouldn't be the recommended way to build.

The first one, let's say `local.Dockerfile` would copy the `setup/` directory that has been build with `ant`. And the second one, `prod.Dockerfile`, would download a release from GitHub. In my opinion both of the resulting images should be based on existing Java images (Zulu, Temurin, etc.).

Both of them should be templated to allow for different base image tags (jre vs jdk).

mgaffigan replied:
Assuming we can't release from an OCI image, then I agree with having a separate dockerfile for running golden bits.  This dockerfile is specifically targeted at developer run, or prod run assuming that we're building inside docker.

mgaffigan replied:
Resolving since https://github.com/OpenIntegrationEngine/engine-docker/pull/1 merged.  Please reopen if there is something unaddressed.

kpalang commented on Dockerfile:
```diff
@@ -0,0 +1,98 @@
+# syntax=docker/dockerfile:1.7-labs
```
suggestion: I think we should stay on `docker/dockerfile:1`

mgaffigan replied:
`docker/dockerfile:1` doesn't support `COPY --exclude=...` and would thus result in increased image sizes.

kpalang replied:
question: Could the `--exclude` directives be run as `rm`-s in the `builder` stage?

mgaffigan replied:
You cannot run an `rm` in the same layer as a COPY, so the image size would still be inflated.  I think the alternative here would be to add an option to `mirth-server.xml` to not copy them to `setup`.

The other options I'm disagreeing with are:
1. Copy each item from setup by name (verbose, fragile if new item added)
1. Copy to a third location (symlink?), then copy to final (ew for complexity)
1. Take the hit on layer-size (48 MB)
1. ADD with special tar (ew for time/space and general disgust)

I generally agree with the advice to stay on `docker/dockerfile:1`, but `--exclude` is fairly well supported by this point, and *should* get [taken out of labs soon](https://github.com/moby/buildkit/issues/2853#issuecomment-2855262799).

mgaffigan replied:
Stages are separate unless a subsequent stage references a prior stage in a "FROM". 

None of the builder stage ends up in the deployment artifact, except that which is specifically COPY'd over.

kpalang commented on DEVELOPERS.md:
````diff
@@ -0,0 +1,34 @@
+# For developers and contributors
+
+## "Easy Path" with docker
+
+```bash
+# Build using docker
+docker build -t oie-dev .
+# Start an ephemeral image
+# NOTE: All data will be deleted on stop due to --rm.  Use a volume for "real" use.
+docker run --rm -p 8443:8443 oie-dev
+```
````
issue: I am against recommending building in Docker as the first approach. Layer caching will become an issue quite soon.

There are two ways to break layer caching:
1. Editing `Dockerfile` to invalidate cache and pull in the newly build directory
    - Way too tedious to be considered acceptable
2. Setting `--no-cache` when building the image
    - Forces rebuilding of _every_ layer and costing unnecessary time

I think the recommended approach should be to use SDKMAN to get the dependencies and leave building containers to people who know what they're doing.

mgaffigan replied:
I'm not sure I follow your comment about layer caching.  That's a feature - not a bug.  `sdk env install` will be re-run any time `.sdkmanrc` is changed.  What cached resource would we not want to keep on build?

kpalang replied:
What I mean was that layer-caching is good on some scenarios, but not necessarily others. What I fear is going to happen is changes in engine code do not trigger a cache miss in the `COPY --exclude=docker . .` layer.

I am testing this right now.

kpalang replied:
Looks like changes in code do indeed trigger a cache miss.

kpalang commented on DEVELOPERS.md:
```diff
@@ -0,0 +1,34 @@
+## Build Environment
+
+To build the solution, you must have a Java 1.8 JDK+FX and Apache Ant.  This
+can be installed by [sdkman](https://sdkman.io/) by executing `sdkman env install`.
```
question: Isn't the command just `sdk` instead of `sdkman`?

mgaffigan replied:
It is.  Updated.

kpalang commented on DEVELOPERS.md:
```diff
@@ -0,0 +1,34 @@
+## Run
+
+After build, run the server by invoking `server/mirth-server-launcher.jar`.  An
+example of how to do this is listed in `docker/mirth-connect.sh`.
```
issue: Now non-existant reference.

I assume the reference was correct in the previous, larger version of the PR.

mgaffigan replied:
Yep.  Updated.

mgaffigan commented on connectservice/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM node:18
```
Accidental merge - does not belong in this PR. I'll take this comment as feedback for https://github.com/OpenIntegrationEngine/engine/pull/114.

mgaffigan commented on connectservice/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+WORKDIR /app
+
+COPY package*.json ./
+RUN npm install
```
Accidental merge - does not belong in this PR. I'll take this comment as feedback for https://github.com/OpenIntegrationEngine/engine/pull/114.

mgaffigan commented on .dockerignore:
Since we are building from source, I think an exclude list would be substantially more manageable than an include list.

kpalang replied:
When writing the previous comment, I was thinking of the release container.
But, if we're talking about a build container for development, I still think includelisting the 6 directories + possibly `.sdkmanrc` might be the cleaner way

mgaffigan replied:
Can you help me understand what you are intending?  The primary goal of a `.dockerignore` is to avoid copying build artifacts, not to select things for the docker build.  It should generally match .gitignore (since if you are not putting it in code-control, it probably should not be used for build).

mgaffigan replied:
This PR is exclusively addressing the containerized build of OIE.  I think the engine-docker repo already addresses the creation of a container from artifacts produced outside of docker.

This should be the logical dual of .gitignore.  If it is in git, it needs to be in the docker context.

mgaffigan commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,27 @@
+      context: ../
+      dockerfile: Dockerfile
+      target: jre-run
+      x-bake:
```
This is removed in the latest rev of the PR.  I'm assuming this will land in 126 for a release that pull signed binaries.

mgaffigan commented on Dockerfile:
```diff
@@ -0,0 +1,87 @@
+COPY .sdkmanrc .
+RUN apt-get update\
+    && apt-get install -y zip curl\
+    && curl -s "https://get.sdkman.io?ci=true" | bash \
```
See https://github.com/mgaffigan/oie/commit/123bbe24f14cb132dd1474d11635fb0213b1da43 for changes to avoid sdkman.  Problems are:

- Zulu nor Temurin seem to publish a JDK with FX included.  Install from deb or tgz requires quite a few deps.
- Ant requires separate installation
- Docker revisions can get out of sync with `.sdkmanrc`

I don't love sdkman, but it does seem simpler to use it in the builder container.

kpalang replied:
I agree with the opinion that neither of the two options are perfect.  Using a jdk image can easily drift with our preferred version from `.sdkmanrc`, and installing sdkman in a container is just nasty.

But in this case installing sdkman in the container is the better way to go to ensure as identical as possible envs.

>Zulu nor Temurin seem to publish a JDK with FX included. Install from deb or tgz requires quite a few deps.

I'm not sure this is true. the `8.0.442.fx-zulu` specified in `.sdkmanrc` is one example of that. Unless you're talking about Docker images.

mgaffigan replied:
@kpalang, is this a blocker for you?  I don't see an easier way here.  See the other branch for the alternative.

kpalang commented on Dockerfile:
```diff
@@ -0,0 +1,96 @@
+EXPOSE 8443
+
+USER engine
+ENTRYPOINT ["./configure-from-env.sh"]
+CMD ["./oieserver"]
+
+##########################################
+#
+#     Alpine JRE Image
+#
+##########################################
+
+FROM eclipse-temurin:21.0.7_6-jre-alpine as jre-run
+
+# Alpine does not include bash by default, so we install it
+RUN apk add --no-cache bash
+# useradd and groupadd are not available in Alpine
+RUN addgroup -S engine \
+    && adduser -S -g engine engine \
+    && mkdir -p /opt/engine/appdata \
+    && chown -R engine:engine /opt/engine
+
+WORKDIR /opt/engine
+COPY --chown=engine:engine --from=builder \
+    --exclude=cli-lib \
+    --exclude=mirth-cli-launcher.jar \
+    --exclude=mccommand \
+    --exclude=manager-lib \
+    --exclude=mirth-manager-launcher.jar \
+    --exclude=mcmanager \
+    /app/server/setup ./
+
+VOLUME /opt/engine/appdata
+VOLUME /opt/engine/custom-extensions
+
+EXPOSE 8443
+
+USER engine
+ENTRYPOINT ["./configure-from-env.sh"]
+CMD ["./oieserver"]
```
question: These images have been moved to `engine-docker` and will be built with https://github.com/OpenIntegrationEngine/engine-docker/pull/1. Would it be reasonable to edit the `Dockerfile` in this repo to be a build/local container only?

mgaffigan replied:
Can you clarify what you are suggesting be removed?  I think it is useful to be able to do a local build of ubuntu or alpine, since those are the release environments and would thus both be needed for testing.

kpalang commented on server/basedir-includes/configure-from-env.sh:
issue: @OpenIntegrationEngine/maintainers correct me if I'm wrong, but this file might live in [engine-docker](https://github.com/OpenIntegrationEngine/engine-docker) repo instead. I don't have a good suggestion on how to sync the two.

kpalang commented on Dockerfile:
```diff
@@ -0,0 +1,96 @@
+COPY .sdkmanrc .
+RUN apt-get update\
+    && apt-get install -y zip curl\
+    && curl -s "https://get.sdkman.io?ci=true" | bash \
```
nitpick: Downloading and executing arbitrary scripts is not good practice. The correct way should be to download the binaries from GH and set them up ourselves, but it quite big hassle and I've never completed it, so might be good enough for now.

mgaffigan replied:
I'm not sure I see the difference between using sdkman and installing sdkman - since they both download and run untrusted scripts.

If we were aiming for zero-trust, then we would need to shift to installing from tgz and add hash verifications.  sdkman is fundamentally opposed to that.  Practically, I would recommend separately maintaining and publishing a javafx container then basing this file on that.

kpalang replied:
In my mind the difference is that we have no control over the script being downloaded from `https://get.sdkman.io`. It carries the risks of DNS poisoning (although highly unlikely), and malicious changes by sdkman maintainers themselves - also highly unlikely given their size. But then again, `xz` was even larger.

What I'm saying is this is something we should keep in mind. I don't have a viable alternative right now.

mgaffigan committed:
Add docker build in tree

Adds docker build and configure-from-env script to configure OIE server
based on environment variables at container startup time.

Issue: https://github.com/OpenIntegrationEngine/engine/issues/40
Signed-off-by: Mitch Gaffigan <mitch.gaffigan@comcast.net>

kpalang commented on Dockerfile:
```diff
@@ -0,0 +1,96 @@
+# syntax=docker/dockerfile:1
```
praise: Love the syntax inclusion here

jonbartels replied:
What does syntax do?

mgaffigan replied:
@jonbartels, It sets the minimum required version of docker to interpret the file.  The `--exclude` argument for the copy command was only added in a later syntax.

jonbartels commented on Dockerfile:
```diff
@@ -0,0 +1,96 @@
+#      1b. Build the application
+# 2. Runner Stage: Creates a lightweight image that runs the application using the JRE.
+
+FROM ubuntu:noble-20251013 AS builder
```
Why not `latest` - this is now 2 months out of date

mgaffigan replied:
`ubuntu:noble-20251013` is current ubuntu LTS release.  Specifying a specific tag is considered best practice for reproducible builds (though admittedly that would be a hash, not a tag, when taken to heart).  Dependabot can submit PR's automatically to track latest, which keeps approval in the hands of the project.

I am unopposed to `latest` if preferred.

mgaffigan commented on DEVELOPERS.md:
```diff
@@ -0,0 +1,30 @@
+# For developers and contributors
+
+## Build and run locally
```
This dockerfile contains a build stage, unless you are referring to a dev container.  I do think adding a dev container would be nice, but I haven't done that.

This PR:
1. From any docker host, will build the repo using sdkman and ant
2. Will copy the build artifacts to a new ubuntu or alpine stage, that can be published as an OCI artifact or ran locally

mgaffigan replied:
See https://github.com/mgaffigan/oie/tree/feature/add-devcontainer for a devcontainer.  Works fine locally, but for some reason archive.apache.org is blocking the github codespaces IP Addresses.

mgaffigan commented on DEVELOPERS.md:
```diff
@@ -0,0 +1,30 @@
+## Build and run locally
+
+To build the solution, you must have a Java 1.8 JDK+FX and Apache Ant.  This
+can be installed by [sdkman](https://sdkman.io/) by executing `sdk env install`.
```
Updated language to address windows.

tonygermano commented on Dockerfile:
```diff
@@ -0,0 +1,96 @@
+# syntax=docker/dockerfile:1
+
```
chore: include license header

tonygermano reviewed (changes requested):
Neither of these comments are blocking:

1. This does not work in my environment (podman installed from package manager in debian bookworm) because of the `COPY --exclude` syntax. If I was using docker instead, I think it still would not work without some modification to my environment to tell it to use BuildKit.
2.  At some point the content of `DEVELOPERS.md` will need to be reconciled with the changes to `CONTRIBUTING.md` in #201 , and likely most of this "how to build" documentation should be moved to https://github.com/OpenIntegrationEngine/docs-website where it can be better organized rather than living in this repo.

mgaffigan commented on DEVELOPERS.md:
```diff
@@ -0,0 +1,35 @@
+
+If you are using Mirth Connect Administrator Launcher, you may need to omit
+`-DdisableSigning=true` to support JWS signatures and run MCAL passing `-k -d`
+to make it ignore self-signed certificates. Launchers like 
```
Per discussion in Github, I updated to refer to the docs site, and only kept a reference to Ballista.

mgaffigan commented:
@tonygermano 
> 1. This does not work in my environment (podman installed from package manager in debian bookworm) because of the `COPY --exclude` syntax. If I was using docker instead, I think it still would not work without some modification to my environment to tell it to use BuildKit.

I updated the syntax line to be more specific - hopefully that will give a better error in podman if --exclude is not supported.

> 2. At some point the content of `DEVELOPERS.md` will need to be reconciled with the changes to `CONTRIBUTING.md` in #201 , and likely most of this "how to build" documentation should be moved to https://github.com/OpenIntegrationEngine/docs-website where it can be better organized rather than living in this repo.

Agreed.  Updated DEVELOPERS.md to reference the docs website.  Not sure about reconciling 201. I wouldn't mind moving that reconciliation to #201, since I'm trying for a small PR here that delivers the dockerfile.

tonygermano reviewed (changes requested):
I preferred the short and sweet DEVELOPERS.md before Jon's changes. I don't think we're at the point yet to say the docker build is the preferred method. As I mentioned in my previous review, it doesn't work in my environment, and I plan to continue building without docker, but I wasn't going to block others from using it. I also expect the docker build to be slightly slower because it has to copy the source files into the container first and otherwise is running the same build. The local build will definitely be faster if we get incremental builds working eventually.

I think it's good to let the repo have just enough info and refer to the docs site, which can evolve more easily as things change.

mgaffigan commented:
I've removed developers.md entirely to leave that to a separate PR. While we opine about documentation, this PR nears its 1st birthday.  Perfect is the enemy of done.

This PR being merged is blocking for smoke testing in github pipelines, and frankly for many modern dev workflows.

mgaffigan committed:
Add github to dockerignore, fix ubuntu docker target

Signed-off-by: Mitch Gaffigan <mitch.gaffigan@comcast.net>

jonbartels reviewed (approved):
I have recently worked on Dockerfiles for my own deploys. This update covers updates to the entrypoint script that I have seen in the wild. The new dockerfile seems correct and applies some recent best practices
